### PR TITLE
Refactor data collection.

### PR DESCRIPTION
**Pull request by Heimdell**
## Description

Problem: data collection is a function that calls itself 5 screens below the place it starts and is a pile of different, but unrelated actions, performed simultaneously.

Solution: divide the pile into parts and then combine parts externally.

### Concrete changes

1) The `xrefcheck: ignore X` functionality is implemented as a separate function.
3) The data collection is a fold over the tree, and other reducers can be connected.

### Results

`Anchor`/`Reference` colelctor fits into one screen now.

### The Visitor API

```haskell
foldNode :: (Monoid a, Monad m) => (Node -> m a) -> Node -> m a
```
Runs the provided action, collects the results.

## Related issue(s)

None, preparing for #64 

## :white_check_mark: Checklist for your Pull Request

Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](/README.md)
    - Haddock

- Public contracts
  - [x] Any modifications of public contracts comply with the [Evolution
  of Public Contracts](https://www.notion.so/serokell/Evolution-of-Public-Contracts-2a3bf7971abe4806a24f63c84e7076c5) policy.
  - [x] I added an entry to the [changelog](/CHANGES.md) if my changes are visible to the users
        and
  - [x] provided a migration guide for breaking changes if possible

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [ ] My code complies with the [style guide](docs/code-style.md).

### CODE REVIEW

File: src/Xrefcheck/Scanners/Markdown.hs
Comment (worm2fed) [inline src/Xrefcheck/Scanners/Markdown.hs:26]: This is not accepted yet) but I do not mind
Can you 'vote' here?
https://github.com/serokell/style/pull/11

File: src/Xrefcheck/Scanners/Markdown.hs
Comment (worm2fed) [inline src/Xrefcheck/Scanners/Markdown.hs]: I don't like this silent change. This commit is about removing mixing, but here you commented function for unknown reason

File: exec/Main.hs
Comment (worm2fed) [inline exec/Main.hs:10]: I can see that you make such a changes in your first commit and at the second one
Please, make smth with it - move this changes from first commit to this one or maybe it'll be ok to squash this two commits (but in this case you need update commit title - and provide commit body! with description of your changes)

File: src/Xrefcheck/Scanners/Markdown/ErrorMessage.hs
Comment (worm2fed) [inline src/Xrefcheck/Scanners/Markdown/ErrorMessage.hs:1]: it's 2021 now))
same in other headers you've added

File: tests/Main.hs
Comment (worm2fed) [inline tests/Main.hs:6]: This changes supposed to be at the first commit...

**Review by worm2fed — CHANGES_REQUESTED** [review]
Great PR description! but so dirty commit history..
Can you do something with it? It's become really hard to review your changes.

1. Go through your commits and decide which small steps to your final result this commits lead you.
2. Add description for each commit (as described in our policy) so we can review this steps separately and understand it.
3. Use `git commit --fixup HASH` to fix smth in your previous commits, so we can understand to which commits related your fixes.

File: src/Xrefcheck/Scanners/Markdown.hs
Comment (Heimdell) [reply src/Xrefcheck/Scanners/Markdown.hs:26]: I use `ImportQualifiedPost` that mitigates the problem anyway.

File: src/Xrefcheck/Scanners/Markdown/ErrorMessage.hs
Comment (Martoon-00) [inline src/Xrefcheck/Scanners/Markdown/ErrorMessage.hs:37]: I think this is not a good thing actually. Putting error messages far from the use site, we encourage their divergence from  the actual problem (e.g. the actual error can obtain one more condition when it fires, but it is now easy to forget to update the error message respectively).

I think error messages and where they are used should fit into screen, maybe but in `where`/`letl` block if found to make the code less readable.

**Review by Martoon-00 — COMMENTED** [review]
My opinion, per change:

1. Using post-imports - kaef (though I wonder, didn't @worm2fed implemented this already somewhere else, taking https://github.com/serokell/style/pull/11 into account?)
1.5. Mixins removal - good; `import Universum` addition - ok.
2. Using `polysemy` - I would retract from doing this change until it actually appears necessary, because using it makes things more complex for newcomers, and there might be cases not well supported by it (as with any advanced framework).
3. Using the Visitor pattern - the same thing, the current Markdown scanner already does most of the things we ever want from it and it just works. After the rewrite, we use only `inside` event, so things just got more complex without immediate benefit. And it's not clear whether future changes in the scanner logic, if any, will fit into the current model without increasing its complexity in directions other than that you support now.
4. Error messages extraction to a separate module - responded in another comment. The fact of extraction is good, but error message and condition for it to fire should locate nearby IMO (they didn't before, but putting to a separate module sounds too harsh).
5. Global `-wno-orphans` - this sounds like a coolstory, I think orhpans are not a very good pattern so discouraging it with a warning is actually good. Or not? Also I heard that compilation time slightly increases with each module that allows orphan instances.
6. Trailing `--` in haddock - why not just use the same style as in the remainder of the project :pepe_stare:  If everyone comes with its own preferred style (that also probably changes every several years), that would be a mess.
7. The separation of ignored removal and info gathering itself - really nice to obtain it :+1: 
8. Moving functions from `where` to top-level - neat.

Well, quite a wall of comments, but this PR is sorta big too )

I would extract 1 & 1.5, maybe 4 (changed), 7 and 8 to a separate PR and leave other changes pending for now.

**Comment by worm2fed** [comment]
@Martoon-00 
> Using post-imports - kaef (though I wonder, didn't @worm2fed implemented this already somewhere else, taking Add requirement for qualified imports style#11 into account?)

We didn't merge it yet. I'll update my PR soon!)

**Comment by Martoon-00** [comment]
> @Martoon-00
> 
> > Using post-imports - kaef (though I wonder, didn't @worm2fed implemented this already somewhere else, taking Add requirement for qualified imports style#11 into account?)
> 
> We didn't merge it yet. I'll update my PR soon!)

:+1: 

Just coordinate with @Heimdell who switches this entire project to post-qualified-imports.

File: package.yaml
Comment (dcastro) [inline package.yaml:35]: Now stylish-haskell automatically rewrites `import A qualified as B` to `import qualified A as B`.

You can verify this behaviour by manually running this:

```
stylish-haskell --inplace src/Xrefcheck/Scanners/Markdown.hs
```

If we wanna use `ImportQualifiedPost`, we need to configure stylish-haskell to keep the "ImportQualifiedPost" format.

File: package.yaml
Comment (dcastro) [reply package.yaml:30]: Let's keep `DerivingStrategies`. In fact, we should always explicitly specify which strategy we're using. Implicit strategies are [a bad practice](https://kowainik.github.io/posts/deriving#anyclass-ambiguity).

Comment (dcastro) [reply package.yaml:30]: We should use `-Weverything` (like we do in morley) or, at the very least, `-Wmissing-deriving-strategies` to force us to specify a deriving strategy. Let's do this in a separate PR. @Heimdell can you please create an issue for this?

File: src/Xrefcheck/Scanners/Markdown.hs
Comment (dcastro) [inline src/Xrefcheck/Scanners/Markdown.hs:149]: Personal suggestion: I think using the monomorphic `DList.empty` and `DList.singleton` (in this branch and the others) makes the code easier to read than the polymorphic `mempty` and `pure`.

Unnecessary polymorphism can sometimes hinder readability.

```suggestion
          return $ FileInfoDiff DList.empty $ DList.singleton $ Anchor {aType, aName, aPos}
```

This is arguable and subjective, so I'll leave this up to you :smile:

File: package.yaml
Comment (Heimdell) [reply package.yaml:35]: image

Google finds 0 results though. I'll try to do it, but can't guarantee it is possible.

Comment (Heimdell) [reply package.yaml:35]: Disregard that, it is possible https://github.com/akrmn/stylish-haskell/commit/c5c7b50445e930389a22e303ee0d84679af8a93e#diff-0924d0340532d02b8c3bacd84f133f5c6da1bf2494e872247f3abfc9e1c8d4b1

**Review by dcastro — APPROVED** [review]

**Comment by dcastro** [comment]
@Martoon-00 do you wanna have another look at this before merging, or are you happy enough with merging as-is?

**Pull request merged**
